File: inference-rs/npm/lib/format.js
// The `list` and `models` view formatters — a byte-for-byte port of the Rust CLI's
// rendering (crates/fxtranslate-cli/src/cli.rs). Presentation is a shell concern
// (duplicated on purpose, per notes/13), but pinned against the Rust oracle: every
// column width, size unit, sort order, and trailer must match. The pure decisions
// these render (catalog classification, version-gated pairs) come from the wasm
// core, not from here.

const { displayName } = require("./lang");

/** The production hub. Mirrors `route::PREFERRED_HUB`. */
const PREFERRED_HUB = "en";

/**
 * Unicode scalar (code-point) count of `s` — what Rust's `chars().count()` and
 * `{:<width$}` measure. `String.length` counts UTF-16 units, so a spread is used
 * to count code points instead (matters for any astral char; harmless otherwise).
 *
 * @param {string} s
 * @returns {number}
 */
function scalarLen(s) {
  return [...s].length;
}

/**
 * Left-justify (pad right) `s` to `width` scalars — Rust `{:<width$}`.
 *
 * @param {string} s
 * @param {number} width
 * @returns {string}
 */
function padEnd(s, width) {
  const n = scalarLen(s);
  return n >= width ? s : s + " ".repeat(width - n);
}

/**
 * Right-justify (pad left) `s` to `width` scalars — Rust `{:>width$}`.
 *
 * @param {string} s
 * @param {number} width
 * @returns {string}
 */
function padStart(s, width) {
  const n = scalarLen(s);
  return n >= width ? s : " ".repeat(width - n) + s;
}

/**
 * `{:.1}`-style rounding matching Rust's float formatter (round half to even), so
 * the one-decimal KiB/MiB/GiB sizes agree byte-for-byte with the oracle.
 *
 * @param {number} value
 * @returns {string}
 */
function oneDecimal(value) {
  const scaled = value * 10;
  let rounded = Math.round(scaled);
  // Math.round rounds .5 up; Rust rounds half to even. Correct the exact-tie case.
  if (Math.abs(scaled - Math.trunc(scaled) - 0.5) < Number.EPSILON) {
    const floor = Math.floor(scaled);
    rounded = floor % 2 === 0 ? floor : floor + 1;
  }
  return (rounded / 10).toFixed(1);
}

/**
 * Human-readable byte count on the base-1024 scale — raw bytes under 1 KiB, then
 * one decimal of KiB/MiB/GiB. Mirrors Rust's `human_bytes`.
 *
 * @param {number} n
 * @returns {string}
 */
function humanBytes(n) {
  const KIB = 1024;
  const MIB = KIB * 1024;
  const GIB = MIB * 1024;
  const f = n;
  if (f >= GIB) {
    return `${oneDecimal(f / GIB)} GiB`;
  }
  if (f >= MIB) {
    return `${oneDecimal(f / MIB)} MiB`;
  }
  if (f >= KIB) {
    return `${oneDecimal(f / KIB)} KiB`;
  }
  return `${n} B`;
}

/**
 * ANSI palette `(cyan, green, dim, reset)`, or empty strings when color is off.
 * Mirrors Rust's `palette`.
 *
 * @param {boolean} color
 * @returns {[string, string, string, string]}
 */
function palette(color) {
  return color
    ? ["\x1b[36m", "\x1b[32m", "\x1b[2m", "\x1b[0m"]
    : ["", "", "", ""];
}

/**
 * Whether a `list` language view surfaces `lang` for `query`. Mirrors Rust's
 * `language_query_matches` (shell logic in cli.rs).
 *
 * @param {string} lang
 * @param {string} query
 * @returns {boolean}
 */
function languageQueryMatches(lang, query) {
  const dash = query.indexOf("-");
  if (dash >= 0) {
    const a = query.slice(0, dash);
    const b = query.slice(dash + 1);
    return lang.startsWith(a) || lang.startsWith(b);
  }
  return lang.startsWith(query);
}

/**
 * Whether a `list --all` query selects the pair `src → trg`. Mirrors Rust's
 * `remote::language_matches` — a shallow prefix match; the version gate that makes
 * this list correct lives in the wasm `modelPairs`.
 *
 * @param {string} src
 * @param {string} trg
 * @param {string} query
 * @returns {boolean}
 */
function languageMatches(src, trg, query) {
  const dash = query.indexOf("-");
  if (dash >= 0) {
    const qSrc = query.slice(0, dash);
    const qTrg = query.slice(dash + 1);
    return src.startsWith(qSrc) && trg.startsWith(qTrg);
  }
  return src.startsWith(query) || trg.startsWith(query);
}

/**
 * Render the aligned `src → trg` table for `rows` — the `--all` view. Names and the
 * source tag are padded before color-wrapping. Mirrors Rust's `render_pairs`.
 *
 * @param {[string, string][]} rows
 * @param {boolean} color
 * @param {(s: string) => void} out
 */
function renderPairs(rows, color, out) {
  const wSrc = Math.max(0, ...rows.map(([s]) => scalarLen(displayName(s))));
  const wStag = Math.max(0, ...rows.map(([s]) => scalarLen(s) + 2));
  const wTrg = Math.max(0, ...rows.map(([, t]) => scalarLen(displayName(t))));
  const [cyan, green, dim, reset] = palette(color);
  for (const [s, t] of rows) {
    const sname = padEnd(displayName(s), wSrc);
    const stag = padEnd(`(${s})`, wStag);
    const tname = padEnd(displayName(t), wTrg);
    out(
      `${cyan}${sname}${reset} ${dim}${stag}${reset} ${dim}→${reset} ${green}${tname}${reset} ${dim}(${t})${reset}\n`,
    );
  }
}

/**
 * Render single-direction rows as `source → target (src trg)`. Mirrors Rust's
 * `render_single_direction`.
 *
 * @param {[string, string][]} rows
 * @param {boolean} color
 * @param {(s: string) => void} out
 */
function renderSingleDirection(rows, color, out) {
  const wSrc = Math.max(0, ...rows.map(([s]) => scalarLen(displayName(s))));
  const wTrg = Math.max(0, ...rows.map(([, t]) => scalarLen(displayName(t))));
  const [cyan, green, dim, reset] = palette(color);
  for (const [s, t] of rows) {
    const sname = padEnd(displayName(s), wSrc);
    const tname = padEnd(displayName(t), wTrg);
    out(
      `${cyan}${sname}${reset} ${dim}→${reset} ${green}${tname}${reset} ${dim}(${s} ${t})${reset}\n`,
    );
  }
}

/**
 * The `list --all` view from the version-gated model pairs. Filters by `query`
 * (via {@link languageMatches}) and renders the raw `src → trg` table. Returns the
 * number of pairs shown, or throws with the Rust no-match message.
 *
 * @param {[string, string][]} allPairs  version-gated pairs from wasm `modelPairs`
 * @param {string | undefined} query
 * @param {boolean} color
 * @param {(s: string) => void} out
 * @returns {number}
 */
function writePairs(allPairs, query, color, out) {
  const shown = allPairs.filter(([s, t]) =>
    query === undefined ? true : languageMatches(s, t, query),
  );
  if (shown.length === 0) {
    throw new Error(
      `no model pairs match \`${query ?? ""}\` (${allPairs.length} pairs available; try \`fxtranslate list --all\`)`,
    );
  }
  renderPairs(shown, color, out);
  return shown.length;
}

/**
 * The default `list` view: languages, not raw pairs. Renders the "Fully supported"
 * and "Single-direction models" sections from the wasm catalog. Returns the number
 * of languages shown, or throws with the Rust no-match message. Mirrors Rust's
 * `write_languages`.
 *
 * @param {{ bidirectional: string[], sourceOnly: string[], targetOnly: string[] }} cat
 * @param {string | undefined} query
 * @param {boolean} color
 * @param {(s: string) => void} out
 * @returns {number}
 */
function writeLanguages(cat, query, color, out) {
  /** @param {string} l */
  const keep = (l) => (query === undefined ? true : languageQueryMatches(l, query));
  const bidi = cat.bidirectional.filter(keep);
  const sourceOnly = cat.sourceOnly.filter(keep);
  const targetOnly = cat.targetOnly.filter(keep);
  const total = bidi.length + sourceOnly.length + targetOnly.length;
  if (total === 0) {
    throw new Error(
      `no languages match \`${query ?? ""}\` (try \`fxtranslate list\`, or \`list --all\` for raw pairs)`,
    );
  }

  const [cyan, , dim, reset] = palette(color);

  if (bidi.length > 0) {
    out("Fully supported (translate to and from any other):\n");
    const w = Math.max(0, ...bidi.map((l) => scalarLen(displayName(l))));
    for (const l of bidi) {
      const name = padEnd(displayName(l), w);
      out(`  ${cyan}${name}${reset} ${dim}(${l})${reset}\n`);
    }
  }

  // Single-direction: hub→L (target-only) then L→hub (source-only).
  /** @type {[string, string][]} */
  const oneWay = [
    ...targetOnly.map((l) => /** @type {[string, string]} */ ([PREFERRED_HUB, l])),
    ...sourceOnly.map((l) => /** @type {[string, string]} */ ([l, PREFERRED_HUB])),
  ];
  if (oneWay.length > 0) {
    if (bidi.length > 0) {
      out("\n");
    }
    out("Single-direction models:\n");
    renderSingleDirection(oneWay, color, out);
  }

  return total;
}

/**
 * A cached pair's `Source → Target` label, or `null` when the directory name
 * doesn't carry the hub on one side. Mirrors Rust's `pretty_pair`.
 *
 * @param {string} name
 * @returns {[string, string] | null}
 */
function prettyPair(name) {
  const hub = PREFERRED_HUB;
  if (name.startsWith(`${hub}-`)) {
    return [displayName(hub), displayName(name.slice(hub.length + 1))];
  }
  if (name.endsWith(`-${hub}`)) {
    return [displayName(name.slice(0, name.length - hub.length - 1)), displayName(hub)];
  }
  return null;
}

module.exports = {
  PREFERRED_HUB,
  scalarLen,
  padEnd,
  padStart,
  humanBytes,
  palette,
  languageQueryMatches,
  languageMatches,
  writePairs,
  writeLanguages,
  prettyPair,
};

File: inference-rs/crates/fxtranslate-py/python/fxtranslate/cli/format.py
"""The ``list`` and ``models`` view formatters — a byte-for-byte port of the Rust
CLI's rendering (crates/fxtranslate-cli/src/cli.rs). Presentation is a shell concern
(duplicated on purpose, per notes/13), but pinned against the Rust oracle: every
column width, size unit, sort order, and trailer must match. The pure decisions
these render (catalog classification, version-gated pairs) come from the compiled
``discovery`` surface, not from here.
"""

from typing import Callable, List, Optional, Tuple

from .lang import display_name

# The production hub. Mirrors `route::PREFERRED_HUB`.
PREFERRED_HUB = "en"


def scalar_len(s: str) -> int:
    """Unicode scalar (code-point) count of ``s`` — what Rust's ``chars().count()``
    and ``{:<width$}`` measure. Python ``len`` already counts code points, so this is
    just ``len``; kept named for parity with the Rust/JS intent."""
    return len(s)


def pad_end(s: str, width: int) -> str:
    """Left-justify (pad right) ``s`` to ``width`` scalars — Rust ``{:<width$}``."""
    return s.ljust(width)


def pad_start(s: str, width: int) -> str:
    """Right-justify (pad left) ``s`` to ``width`` scalars — Rust ``{:>width$}``."""
    return s.rjust(width)


def human_bytes(n: int) -> str:
    """Human-readable byte count on the base-1024 scale — raw bytes under 1 KiB, then
    one decimal of KiB/MiB/GiB. Mirrors Rust's ``human_bytes``. Python's ``%.1f``
    rounds half-to-even, matching Rust's float formatter."""
    kib = 1024.0
    mib = kib * 1024.0
    gib = mib * 1024.0
    f = float(n)
    if f >= gib:
        return f"{f / gib:.1f} GiB"
    if f >= mib:
        return f"{f / mib:.1f} MiB"
    if f >= kib:
        return f"{f / kib:.1f} KiB"
    return f"{n} B"


def palette(color: bool) -> Tuple[str, str, str, str]:
    """ANSI palette ``(cyan, green, dim, reset)``, or empty strings when color is off.
    Mirrors Rust's ``palette``."""
    if color:
        return ("\x1b[36m", "\x1b[32m", "\x1b[2m", "\x1b[0m")
    return ("", "", "", "")


def language_query_matches(lang: str, query: str) -> bool:
    """Whether a ``list`` language view surfaces ``lang`` for ``query``. Mirrors Rust's
    ``language_query_matches``."""
    dash = query.find("-")
    if dash >= 0:
        a = query[:dash]
        b = query[dash + 1 :]
        return lang.startswith(a) or lang.startswith(b)
    return lang.startswith(query)


def language_matches(src: str, trg: str, query: str) -> bool:
    """Whether a ``list --all`` query selects the pair ``src → trg``. Mirrors Rust's
    ``remote::language_matches`` — a shallow prefix match; the version gate that makes
    this list correct lives in the compiled ``model_pairs``."""
    dash = query.find("-")
    if dash >= 0:
        q_src = query[:dash]
        q_trg = query[dash + 1 :]
        return src.startswith(q_src) and trg.startswith(q_trg)
    return src.startswith(query) or trg.startswith(query)


def _render_pairs(rows: List[Tuple[str, str]], color: bool, out: Callable[[str], None]) -> None:
    """Render the aligned ``src → trg`` table — the ``--all`` view. Names and the
    source tag are padded before color-wrapping. Mirrors Rust's ``render_pairs``."""
    w_src = max((scalar_len(display_name(s)) for s, _ in rows), default=0)
    w_stag = max((scalar_len(s) + 2 for s, _ in rows), default=0)
    w_trg = max((scalar_len(display_name(t)) for _, t in rows), default=0)
    cyan, green, dim, reset = palette(color)
    for s, t in rows:
        sname = pad_end(display_name(s), w_src)
        stag = pad_end(f"({s})", w_stag)
        tname = pad_end(display_name(t), w_trg)
        out(
            f"{cyan}{sname}{reset} {dim}{stag}{reset} {dim}→{reset} "
            f"{green}{tname}{reset} {dim}({t}){reset}\n"
        )


def _render_single_direction(
    rows: List[Tuple[str, str]], color: bool, out: Callable[[str], None]
) -> None:
    """Render single-direction rows as ``source → target (src trg)``. Mirrors Rust's
    ``render_single_direction``."""
    w_src = max((scalar_len(display_name(s)) for s, _ in rows), default=0)
    w_trg = max((scalar_len(display_name(t)) for _, t in rows), default=0)
    cyan, green, dim, reset = palette(color)
    for s, t in rows:
        sname = pad_end(display_name(s), w_src)
        tname = pad_end(display_name(t), w_trg)
        out(f"{cyan}{sname}{reset} {dim}→{reset} {green}{tname}{reset} {dim}({s} {t}){reset}\n")


def write_pairs(
    all_pairs: List[Tuple[str, str]],
    query: Optional[str],
    color: bool,
    out: Callable[[str], None],
) -> int:
    """The ``list --all`` view from the version-gated model pairs. Filters by ``query``
    (via :func:`language_matches`) and renders the raw ``src → trg`` table. Returns the
    number of pairs shown, or raises with the Rust no-match message."""
    shown = [(s, t) for (s, t) in all_pairs if query is None or language_matches(s, t, query)]
    if not shown:
        raise ValueError(
            f"no model pairs match `{query or ''}` ({len(all_pairs)} pairs available; "
            "try `fxtranslate list --all`)"
        )
    _render_pairs(shown, color, out)
    return len(shown)


def write_languages(
    cat: dict,
    query: Optional[str],
    color: bool,
    out: Callable[[str], None],
) -> int:
    """The default ``list`` view: languages, not raw pairs. Renders the "Fully
    supported" and "Single-direction models" sections from the compiled catalog.
    Returns the number of languages shown, or raises with the Rust no-match message.
    Mirrors Rust's ``write_languages``."""

    def keep(l: str) -> bool:
        return query is None or language_query_matches(l, query)

    bidi = [l for l in cat["bidirectional"] if keep(l)]
    source_only = [l for l in cat["sourceOnly"] if keep(l)]
    target_only = [l for l in cat["targetOnly"] if keep(l)]
    total = len(bidi) + len(source_only) + len(target_only)
    if total == 0:
        raise ValueError(
            f"no languages match `{query or ''}` "
            "(try `fxtranslate list`, or `list --all` for raw pairs)"
        )

    cyan, _green, dim, reset = palette(color)

    if bidi:
        out("Fully supported (translate to and from any other):\n")
        w = max((scalar_len(display_name(l)) for l in bidi), default=0)
        for l in bidi:
            name = pad_end(display_name(l), w)
            out(f"  {cyan}{name}{reset} {dim}({l}){reset}\n")

    # Single-direction: hub→L (target-only) then L→hub (source-only).
    one_way: List[Tuple[str, str]] = [(PREFERRED_HUB, l) for l in target_only] + [
        (l, PREFERRED_HUB) for l in source_only
    ]
    if one_way:
        if bidi:
            out("\n")
        out("Single-direction models:\n")
        _render_single_direction(one_way, color, out)

    return total


def pretty_pair(name: str) -> Optional[Tuple[str, str]]:
    """A cached pair's ``Source → Target`` label, or ``None`` when the directory name
    doesn't carry the hub on one side. Mirrors Rust's ``pretty_pair``."""
    hub = PREFERRED_HUB
    if name.startswith(f"{hub}-"):
        return (display_name(hub), display_name(name[len(hub) + 1 :]))
    if name.endswith(f"-{hub}"):
        return (display_name(name[: len(name) - len(hub) - 1]), display_name(hub))
    return None
